Source/WebCore/page/Navigation.cpp

```diff
@@ -613,6 +614,50 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
     }
 }
 
+static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(Document& document)
+{
+    auto& globalObject = *JSC::jsCast<JSDOMGlobalObject*>(document.globalObject());
+    JSC::JSLockHolder lock(globalObject.vm());
+    RefPtr deferredPromise = DeferredPromise::create(globalObject);
+    Ref domPromise = DOMPromise::create(globalObject, *JSC::jsCast<JSC::JSPromise*>(deferredPromise->promise()));
+    return { WTFMove(domPromise), deferredPromise.releaseNonNull() };
+}
+
+// https://webidl.spec.whatwg.org/#wait-for-all
+static void waitForAllPromises(Vector<RefPtr<DOMPromise>>& promises, std::function<void()> fulfilledCallback, std::function<void(JSC::JSValue)> rejectionCallback)
+{
+    size_t remainingPromises = promises.size();
+    bool rejected = false;
+
+    for (const auto& promise : promises) {
+        // At any point between promises the frame could have been detached.
+        if (promise->isSuspended())
+            return;
+
+        promise->whenSettled([=] () mutable {
```

Comment:
Can you explain how the capturing works correctly in waiting for all promises to resolve? I have read about all capture modes in a C++ book some time ago but I forgot.

Reply:
Oops, this capture wasn't right, I did want to copy the promise but it needs to ref the count.

Reply:
This copies the pointers it references, just adding refs in this case.